File: backend/app/memory/keyword.py
"""Keyword-based retrieval baseline."""

from dataclasses import dataclass

from app.memory.encoder import _tokenize, _tokenize_with_surface


@dataclass
class KeywordResult:
    memory_id: str
    score: float
    matched_terms: list[str]


def keyword_search(query: str, memories: list[dict], top_k: int = 5) -> list[KeywordResult]:
    """Token-overlap retrieval over stemmed tokens; matched_terms reports surface forms."""
    query_pairs = _tokenize_with_surface(query)
    if not query_pairs:
        return []

    query_stems = [stem for stem, _ in query_pairs]
    surface_by_stem = {stem: surface for stem, surface in query_pairs}

    results = []
    for mem in memories:
        doc_tokens: set[str] = set(_tokenize(mem["text"]))
        for e in mem.get("entities", []) or []:
            doc_tokens.update(_tokenize(e))
        for t in mem.get("tags", []) or []:
            doc_tokens.update(_tokenize(t))
        if mem.get("subject"):
            doc_tokens.update(_tokenize(mem["subject"]))
        if mem.get("predicate"):
            doc_tokens.update(_tokenize(mem["predicate"]))
        if mem.get("object"):
            doc_tokens.update(_tokenize(mem["object"]))

        matched_stems = [s for s in query_stems if s in doc_tokens]
        if matched_stems:
            score = len(matched_stems) / len(query_stems)
            surface_matched = [surface_by_stem[s] for s in matched_stems]
            results.append(KeywordResult(
                memory_id=mem["id"],
                score=score,
                matched_terms=surface_matched,
            ))

    results.sort(key=lambda r: r.score, reverse=True)
    return results[:top_k]

File: frontend/lib/hrr/keyword.ts
import { tokenize, tokenizeWithSurface } from "./encoder";

export interface KeywordResult {
  memoryId: string;
  score: number;
  matchedTerms: string[];
}

interface MemoryDoc {
  id: string;
  text: string;
  subject?: string | null;
  predicate?: string | null;
  object?: string | null;
  entities?: string[];
  tags?: string[];
}

export function keywordSearch(query: string, memories: MemoryDoc[], topK: number = 5): KeywordResult[] {
  const queryPairs = tokenizeWithSurface(query);
  if (queryPairs.length === 0) return [];

  const queryStems = queryPairs.map((p) => p.stem);
  const surfaceByStem = new Map(queryPairs.map((p) => [p.stem, p.surface]));

  const results: KeywordResult[] = [];

  for (const mem of memories) {
    const docTokens = new Set<string>();
    for (const t of tokenize(mem.text)) docTokens.add(t);
    if (mem.subject) for (const t of tokenize(mem.subject)) docTokens.add(t);
    if (mem.predicate) for (const t of tokenize(mem.predicate)) docTokens.add(t);
    if (mem.object) for (const t of tokenize(mem.object)) docTokens.add(t);
    if (mem.entities) for (const e of mem.entities) for (const t of tokenize(e)) docTokens.add(t);
    if (mem.tags) for (const t of mem.tags) for (const tt of tokenize(t)) docTokens.add(tt);

    const matchedStems = queryStems.filter((s) => docTokens.has(s));
    if (matchedStems.length > 0) {
      results.push({
        memoryId: mem.id,
        score: matchedStems.length / queryStems.length,
        matchedTerms: matchedStems.map((s) => surfaceByStem.get(s) ?? s),
      });
    }
  }

  results.sort((a, b) => b.score - a.score);
  return results.slice(0, topK);
}
